Add version of UnusedTracker that supports full shrinking using R8.

netomi opened the pull request:
Updated PR for #566 
Adds support for #565

netomi commented:
Changes that I have performed since the last PR:

* Extract UnusedTracker as an abstract base class for different implementations
* Existing code is moved to UnusedTrackerWithJDependency and used by default
* add an useR8 option to the ShadowSpec to enable R8 minimization
* use the latest R8 release available on the google maven repo
* separate unit tests in MinimizeR8Spec
* Implement UnusedTrackerWithR8 in java instead of groovy to avoid groovy madness in combination with R8

I applied all comments apart from the dependency on R8, its still implementation. I wanted to keep it like that to be able to easily write unit tests using the new shrinking mode. However I am perfectly fine to change it to a compileOnly dependency and check if R8 is available at runtime and use it in that case.

JD-CSTx commented:
@netomi Yeah: Your CI tests fail for the same reason as in my pull request: #773. Link to comment: https://github.com/johnrengelman/shadow/pull/773#issuecomment-1268201381.

netomi commented:
> @netomi Yeah: Your CI tests fail for the same reason as in my pull request: #773. Link to comment: [#773 (comment)](https://github.com/johnrengelman/shadow/pull/773#issuecomment-1268201381).

I have seen that, it comes from this commit:

https://github.com/johnrengelman/shadow/commit/38566d50345de45d25b1404669f9e1d6856e5e99

Seems easy to fix, will create a PR.

JD-CSTx commented:
@netomi I'm not so sure. The original PR author had a valid case for his/her PR: "Do not add a dependencies block if it's already there". But the PR seemed to be incomplete. I saw someone else just reverting it, I don't think that's a good idea, because the idea behind it is valid.

netomi commented:
> @netomi I'm not so sure. The original PR author had a valid case for his/her PR: "Do not add a dependencies block if it's already there". But the PR seemed to be incomplete. I saw someone else just reverting it, I don't think that's a good idea, because the idea behind it is valid.

the change assumes that if a node 'dependencies' exists, it is never empty as it accesses the item at index 0. In the tests you can see that the node is actually empty and thus fails with an index out of bounds exception.

netomi commented:
The build issues wrt publication is fixed in PR #798

JD-CSTx commented:
> The build issues wrt publication is fixed in PR #798

Big thanks.

netomi commented:
Updated the documentation and added possibility to specify additional rules via the shadow spec. I would consider this PR to be finished from my side unless there is some suggestion for change. With the additional rules that can be specified via the shadow spec, this method should allow enough flexibility to handle more complex scenarios.

netomi commented:
A more realistic use case of the minimization with R8 can be tested here: https://github.com/netomi/uom/tree/shadow-dependencies

`./gradlew clean shadowJar`

In this project 2 classes from external dependencies are used (from guava and hibernate). Without minimization or the previous one, a huge number of classes is copied in the shadow jar. With R8 enabled, the copied dependencies is minimal as expected.

gildor commented on src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/DefaultR8Configuration.groovy:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    @Override
+    void rule(String rule) {
```
It would be nice to allow load rules from config file, maybe something like `rules(file("proguard.conf"))`

netomi replied:
Makes sense, will add this.
